File: python/longest_valid_parentheses.py
"""
slow solution:

class Solution:
    def longestValidParentheses(self, s: str) -> int:
        if not len(s): return 0
        paren_depth, min_depth, max_depth = 0, 0, 0
        starting_depths_to_indices = defaultdict(list)
        longest_lengths_by_starting_index = [0 for i in range(len(s))]
        for i, ch in enumerate(s):
            is_open = ch == "("
            if is_open:
                paren_depth += 1
                max_depth = max(paren_depth, max_depth)
                starting_depths_to_indices[paren_depth].append(i)
            else: 
                paren_depth -= 1
                min_depth = min(paren_depth, min_depth)
            for j in range(paren_depth + 2, max_depth + 1):
                starting_depths_to_indices[j] = []
            for index in starting_depths_to_indices[paren_depth + 1]:
                longest_lengths_by_starting_index[index] = i + 1 - index
        return max(longest_lengths_by_starting_index)
            
            
            
"""

class Solution:
    def longestValidParentheses(self, s: str) -> int:
        #search for substrings where depth at start_index is one greater than depth at end_index - 1, 
        #and where no intermediate index has depth less than depth at end_index - 1
        starting_index = None
        starting_depth = None
        cur_depth = 0
        longest_found = 0
        for i, ch in enumerate(s):
            is_open = ch == "("
            cur_depth += 1 if is_open else -1
            #if starting index is None, start a new segment
            if is_open and starting_index is None:
                starting_index, starting_depth = i, cur_depth
            #if the present index satisfies the conditions for the end of a substring beginning with the current starting_index,
            #check if this substring is the longest found so far
            if ( not is_open and starting_depth is not None and 
                starting_depth - 1 == cur_depth ):
                longest_found = max(longest_found, i + 1 - starting_index)
                #if there are no possible valid substrings with the same start_index that end beyond this index, start a new segment
                #by setting starting_index and starting_depth to None 
                if i == len(s) - 1 or s[i + 1] == ")":
                    starting_index, starting_depth = None, None
        return max(longest_found, self.endingValidParentheses(s))
    
    #check from oposite direction
    def endingValidParentheses(self, s: str) -> int:
        #begin checking by locating the last close parentheses
        last_index = s.rfind(")")
        longest_found = 0
        cur_depth = 1
        for i in range(last_index - 1, -1, -1):
            cur_depth += 1 if s[i] == ")" else -1
            #if cur_depth is zero, this is a valid substring
            if cur_depth == 0:
                longest_found = max(longest_found, last_index + 1 - i)
                #if there are no valid substrings that include this substring, check the remaining portion of the string
                if i == 0 or s[i - 1] == "(":
                    return max(longest_found, self.endingValidParentheses(s[:i]))
        return longest_found

File: java/longestValidParentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        //check in both directions and take the max
        String reverse_s = reverseParentheses(s);
        return Math.max(longestForward(s), longestForward(reverse_s));
    }
    
    public int longestForward(String s) {
        //we'll be searching for substrings where the depth at the startIndex is one greater than the depth at the end
        int curDepth = 0;
        //set startIndex to -1 and startDepth to 0 when we're not currently searching a substring
        int startIndex = -1;
        int startDepth = 0;
        int longest_found = 0;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            boolean isOpen = c == '(';
            curDepth += isOpen ? 1 : -1;
            if (isOpen && startIndex == -1) {
                startIndex = i;
                startDepth = curDepth;
            }
            //if the present index forms a valid substring with the present startIndex, 
            //see if this substring is the longest we've found so far
            if (!isOpen && startIndex != -1 && startDepth - 1 == curDepth) {
                longest_found = Math.max(longest_found, i + 1 - startIndex);
                //if there are no more valid substrings with this startIndex, start a new segment by
                //resetting startIndex and startDepth
                if (i == s.length() - 1 || s.charAt(i + 1) == ')') {
                    startIndex = -1;
                    startDepth = 0;
                }
            }
        }
        return longest_found;
    }
    
    public String reverseParentheses(String s) {
        StringBuilder sb = new StringBuilder();
        for (int i = s.length() - 1; i >= 0; i--) {
            char c = s.charAt(i) == ')' ? '(' : ')';
            sb.append(c);
        }
        return sb.toString();
    }
}